### PR TITLE
feat: add IPConfig state logs/metrics to Request/Release handlers

### PR DESCRIPTION
**Pull request by rbtr**
**Reason for Change**:

Logs and records IPAM metrics on Request/Release and MarkPendingRelease calls.

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] relevant PR labels added

**Notes**:

### CODE REVIEW

File: cns/restserver/metrics.go [hunk 1]
```diff
@@ -142,8 +144,51 @@ func stateTransitionMiddleware(i *cns.IPConfigurationStatus, s types.IPState) {
...
+			state.releasingIPs++
+		}
+	}
+
+	logger.Printf("Allocated IPs: %d, Assigned IPs: %d, Available IPs: %d, PendingProgramming IPs: %d, PendingRelease IPs: %d",
+		state.allocatedIPs,
+		state.assignedIPs,
+		state.availableIPs,
+		state.programmingIPs,
+		state.releasingIPs,
```
Comment (timraymond) [inline cns/restserver/metrics.go:188]: Worth it to implement `fmt.Formatter` for `ipState`? I feel like you'd always want to see that output when dumping that struct into a string.

File: go.mod [hunk 2]
```diff
@@ -1,8 +1,6 @@
 module github.com/Azure/azure-container-networking
 
-go 1.22.0
-
-toolchain go1.22.7
+go 1.23
```
Comment (timraymond) [inline go.mod:3]: ❤️

File: cns/restserver/metrics.go [hunk 3]
```diff
@@ -142,8 +144,51 @@ func stateTransitionMiddleware(i *cns.IPConfigurationStatus, s types.IPState) {
...
+	// programmingIPs are the IPs in state "PendingProgramming".
+	programmingIPs int64
+	// releasingIPs are the IPs in state "PendingReleasr".
+	releasingIPs int64
+}
+
+// publishIPStateMetrics logs and publishes the IP Config state metrics to Prometheus.
+func (service *HTTPRestService) publishIPStateMetrics() {
+	// copy state
+	service.RLock()
```
Comment (timraymond) [inline cns/restserver/metrics.go:163]: Could this cause a slow hangup from the server if say:

G1: Finishes handler
G1: defer service.Unlock() runs
G1: publishIPStateMetrics() starts
G2: Starts handler
G2: service.Lock()
G2: sleep for an eternity
G2: defer service.Unlock() runs
G2: publishIPStateMetrics() runs to completion
G1: service.RLock()
G1: finishes publishIPStateMetrics() and releases locks
G1: writes 200 and closes socket now that the handler func is finally done

File: cns/restserver/metrics.go (hunk 1 above)
Comment (rbtr) [reply cns/restserver/metrics.go:188]: yes, but it shouldn't be dumped to a string anywhere else ever... is it worth it for private throwaway containers?

File: cns/restserver/metrics.go (hunk 3 above)
Comment (rbtr) [reply cns/restserver/metrics.go:163]: valid point, need to reconsider this. i hate just shoving more things in to the server Locked scope but the mutexing on this god object is not very precise

File: cns/restserver/metrics.go (hunk 1 above)
Comment (timraymond) [reply cns/restserver/metrics.go:188]: Yeah, probably not in that case. Just thinking about ways things could be different.

File: cns/restserver/metrics.go [hunk 4]
```diff
@@ -142,11 +143,88 @@ func stateTransitionMiddleware(i *cns.IPConfigurationStatus, s types.IPState) {
...
+
+// publishIPStateMetrics logs and publishes the IP Config state metrics to Prometheus.
+func (service *HTTPRestService) publishIPStateMetrics() {
+	if recorder == nil {
+		recorder = &asyncMetricsRecorder{
+			podIPConfigSrc: service.PodIPConfigStates,
+			sig:            make(chan struct{}),
+		}
+		go recorder.run()
+	}
```
Comment (timraymond) [inline cns/restserver/metrics.go:217]: I think the race detector might complain about this. Two goroutines read `recorder`, decide it's nil, then instantiate and assign to `recorder` and start a goroutine. N-1 of those will leak.

Not a huge deal, but `sync.Once` might make sense here.

Comment (rbtr) [reply cns/restserver/metrics.go:217]: I didn't want to add another field to the HTTPRestService for this, what do you think about the once living in the singelton?

Comment (timraymond) [reply cns/restserver/metrics.go]: I think that's probably fine

**Review by timraymond — APPROVED** [review]

**Pull request merged**